File: 382.py
import random


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        ans = 0
        i = 0
        node = self.head
        while node:
            if random.randint(0, i) == 0:
                ans = node.val
            i += 1
            node = node.next
        return ans


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()

File: 382.go
package leetcode

import "math/rand"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type Solution struct {
	head *ListNode
}

func Constructor(head *ListNode) Solution {
	return Solution{
		head,
	}
}

func (this *Solution) GetRandom() int {
	ans := 0
	head := this.head
	for i := 0; head != nil; i++ {
		if rand.Intn(i+1) == 0 { // 1/i的概率被选中
			ans = head.Val
		}
		head = head.Next
	}
	return ans
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(head);
 * param_1 := obj.GetRandom();
 */
